src/beautifiers/rubocop.coffee

```diff
@@ -17,52 +18,54 @@ module.exports = class Rubocop extends Beautifier
   }
 
   beautify: (text, language, options) ->
+    editor = atom?.workspace?.getActiveTextEditor()
+    if editor?
+      fullPath = editor.getPath()
+      [projectPath, relativePath] = atom.project.relativizePath(fullPath)
+    else
+      throw new Error("No active editor found!")
+
+    # Find the rubocop path
+    rubocopPath = "rubocop"
     @Promise.all([
       @which(options.rubocop_path) if options.rubocop_path
       @which('rubocop')
-    ]).then((paths) =>
+    ])
+    .then((paths) =>
       @debug('rubocop paths', paths)
-      _ = require 'lodash'
-      path = require 'path'
       # Get first valid, absolute path
-      rubocopPath = _.find(paths, (p) -> p and path.isAbsolute(p) )
+      rubocopPath = paths.find((p) -> p and path.isAbsolute(p))
       @verbose('rubocopPath', rubocopPath)
       @debug('rubocopPath', rubocopPath, paths)
+    )
 
-      configFile = path.join(atom.project.getPaths()[0], ".rubocop.yml")
-
-      fs = require 'fs'
+    # Find or generate a config file if non exists
+    configFile = @findFile(path.dirname(fullPath), ".rubocop.yml")
+    if !configFile?
+      yaml = require("yaml-front-matter")
+      config = {
+        "Style/IndentationWidth":
+          "Width": options.indent_size
+      }
+      tempConfig = @tempFile("rubocop-config", yaml.safeDump(config))
 
-      if fs.existsSync(configFile)
-        @debug("rubocop", config, fs.readFileSync(configFile, 'utf8'))
-      else
-        yaml = require("yaml-front-matter")
-        # Generate config file
-        config = {
-          "Style/IndentationWidth":
-            "Width": options.indent_size
-        }
+    rubocopArguments = [
+      "--auto-correct"
+      "--force-exclusion"
+      "--stdin", relativePath
+    ]
+    rubocopArguments.push("--config", tempConfig) if tempConfig?
+    @debug("rubocop arguments", rubocopArguments)
 
-        configFile = @tempFile("rubocop-config", yaml.safeDump(config))
-        @debug("rubocop", config, configFile)
+    @run(rubocopPath, rubocopArguments, {
+      ignoreReturnCode: true,
+      cwd: projectPath,
+      onStdin: (stdin) -> stdin.end text
+    }).then((stdout) =>
+      @debug("rubocop output", stdout)
+      # Rubocop output an error if stdout is empty
+      return text if stdout.length == 0
 
-      # Check if PHP-CS-Fixer path was found
-      if rubocopPath?
-        @run(rubocopPath, [
-          "--auto-correct"
-          "--config", configFile
-          tempFile = @tempFile("temp", text, '.rb')
-          ], {ignoreReturnCode: true})
-          .then(=>
-            @readFile(tempFile)
-          )
-      else
-        @run("rubocop", [
-          "--auto-correct"
-          "--config", configFile
-          tempFile = @tempFile("temp", text, '.rb')
-          ], {ignoreReturnCode: true})
-          .then(=>
-            @readFile(tempFile)
-          )
-)
+      result = stdout.split("====================\n")
+      result[result.length - 1]
```

Comment:
What does the `stdout` look like from Rubocop when `--stdin` is enabled?

If Rubocop requires these 2 lines to parse the output, I think I would be more comfortable using the original file output.

Reply:
Unfortunately that would reintroduce the problem this patch is attempting to fix.  Rubocop either 1) uses `--stdin` and outputs to `stdout`, or 2) it changes the file in place. There's no other way to do it. Rubocop has an `--out <file>` switch that you think would write out the file, right? Nope, it writes out the *report*... and overwrites the original file anyway. 😣 

After way too many hours of wrestling with Rubocop just to behave properly, this is the only solution I found that covers all of the use cases and doesn't require any hacks. For better or worse, the way that I'm doing it now is how Rubocop intends it to be integrated into an editor: 

```
    -s, --stdin FILE                 Pipe source from STDIN, using FILE in offense
                                     reports. This is useful for editor integration.
```

To answer your specific question, Rubocop's stdout in this case is the list of offenses, and then the split break, and then the auto corrected file. There is no way to suppress the list of offenses and just output the file.

Here's an example of what the stdout looks like from Rubocop:

```
> cat test.rb | rubocop --stdin test.rb -a
Inspecting 1 file
C

Offenses:

test.rb:1:1: C: Missing top-level class documentation comment.
class Test
^^^^^
test.rb:3:3: C: [Corrected] Use 2 (not 4) spaces for indentation.
      @sample = sample
  ^^^^
test.rb:5:1: C: [Corrected] Extra empty line detected at class body end.
test.rb:5:1: C: [Corrected] Trailing whitespace detected.
test.rb:6:1: C: [Corrected] Extra blank line detected.
test.rb:6:1: C: [Corrected] Extra empty line detected at class body end.
test.rb:6:1: C: [Corrected] Trailing whitespace detected.

1 file inspected, 7 offenses detected, 6 offenses corrected
====================
class Test
  def initialize(sample)
    @sample = sample
  end
end
```

Reply:
Thank you for this in depth analysis. I can now understand these decisions you made, and it makes sense 👍 .

> There is no way to suppress the list of offenses and just output the file.

I took a quick look at some of the other Atom + Rubocop packages and they used a `--format json` argument.
See https://github.com/AtomLinter/linter-rubocop/blob/master/src/index.js#L13 and https://github.com/tomoya/rubocop-auto-correct/blob/master/lib/rubocop-auto-correct.coffee#L94
Could you investigate this option? It may be clearer solution. If not, then we can keep it how it is, although I am concerned about future versions of Rubcop simply changing the separator from `====================` and breaking Atom-Beautify 😝 .

Reply:
I wish! I believe this is as elegant as we can get without hacking on Rubocop. The formatters only output the offenses, but not what would change in the file. This is useful for displaying linter issues in the gutter, but not for actually fixing the files. You *can* combine `--format son` with `--stdin`, but this is even worse for parsing because there's no line break before the separator:

```
... {"offense_count":6,"target_file_count":1,"inspected_file_count":1}}====================
class Test
  def initialize(thing)
    @secret = "my secret"
  end
end
```
😣

Those two tools you listed either also use `--stdin` (linter-rubocop) or allow Rubocop modify the file in place (rubocop-auto-correct). I believe that atom-beautify expects beautifiers *not* to modify files in place, instead returning a buffer of the fixed text. If that's incorrect, we can let Rubocop change the file itself, which would be easier.

Reply:
> I believe that atom-beautify expects beautifiers not to modify files in place, instead returning a buffer of the fixed text. 

You are correct. 

I think your assessment is correct and we are stuck. Keep the current workaround 👍 .